pkg/authorization/decorator_test.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorization
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+
+	auditapis "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+)
+
+func TestDecorator(t *testing.T) {
+	alwaysAllow := authorizer.AuthorizerFunc(func(ctx context.Context, a authorizer.Attributes) (authorizer.Decision, string, error) {
+		return authorizer.DecisionAllow, "unanonymized allow", nil
+	})
+	alwaysDeny := authorizer.AuthorizerFunc(func(ctx context.Context, a authorizer.Attributes) (authorizer.Decision, string, error) {
+		return authorizer.DecisionDeny, "unanonymized denial", nil
+	})
+	alwaysNoOpinion := authorizer.AuthorizerFunc(func(ctx context.Context, a authorizer.Attributes) (authorizer.Decision, string, error) {
+		return authorizer.DecisionNoOpinion, "unanonymized no-opinion", nil
+	})
+	alwaysError := authorizer.AuthorizerFunc(func(ctx context.Context, a authorizer.Attributes) (authorizer.Decision, string, error) {
+		return authorizer.DecisionNoOpinion, "unanonymized failure", errors.New("unanonymized error")
+	})
+
+	for name, tc := range map[string]struct {
+		authz        authorizer.Authorizer
+		wantDecision authorizer.Decision
+		wantAudit    map[string]string
+		wantReason   string
+	}{
+		"topAllows": {
+			authz: NewDecorator("top", alwaysAllow).AddAuditLogging().AddAnonymization().AddReasonAnnotation(),
+
+			wantDecision: authorizer.DecisionAllow,
+			wantReason:   "top: access granted",
+
+			wantAudit: map[string]string{
+				"top/decision": "Allowed",
+				"top/reason":   "unanonymized allow",
```

Comment:
If this is with anonymization, why is the reason unanonymyzed?

Reply:
Similar to `AddReasonAnnotation` anonmyzation refers to what reason bubbles up in the chain, in the audit we want to see the unanomyzied entry, let me add a godoc comment here as well.